src/avro/AvroSchemaEncoder.ts

```diff
@@ -0,0 +1,453 @@
+import type {IWriter, IWriterGrowable} from '@jsonjoy.com/util/lib/buffers';
+import {AvroEncoder} from './AvroEncoder';
+import {AvroSchemaValidator} from './AvroSchemaValidator';
+import type {
+  AvroSchema,
+  AvroRecordSchema,
+  AvroEnumSchema,
+  AvroArraySchema,
+  AvroMapSchema,
+  AvroUnionSchema,
+  AvroFixedSchema,
+  AvroNamedSchema,
+} from './types';
+
+/**
+ * Apache Avro binary encoder with schema validation and encoding.
+ * Encodes values according to provided Avro schemas with proper validation.
+ * Based on https://avro.apache.org/docs/1.12.0/specification/
+ */
+export class AvroSchemaEncoder {
+  private encoder: AvroEncoder;
+  private validator: AvroSchemaValidator;
+  private namedSchemas = new Map<string, AvroNamedSchema>();
+
+  constructor(public readonly writer: IWriter & IWriterGrowable) {
+    this.encoder = new AvroEncoder(writer);
+    this.validator = new AvroSchemaValidator();
+  }
+
+  /**
+   * Encodes a value according to the provided schema.
+   */
+  public encode(value: unknown, schema: AvroSchema, selectedIndex?: number): Uint8Array {
+    this.writer.reset();
+    this.namedSchemas.clear();
+    
+    // Validate schema first
+    if (!this.validator.validateSchema(schema)) {
+      throw new Error('Invalid Avro schema');
+    }
+    
+    // Validate value against schema
+    if (!this.validator.validateValue(value, schema)) {
+      throw new Error('Value does not conform to schema');
+    }
+    
+    this.collectNamedSchemas(schema);
+    
+    if (Array.isArray(schema) && selectedIndex !== undefined) {
+      this.writeUnion(value, schema, selectedIndex);
+    } else {
+      this.writeValue(value, schema);
+    }
+    
+    return this.writer.flush();
+  }
+
+  /**
+   * Writes a null value with schema validation.
+   */
+  public writeNull(schema: AvroSchema): void {
```

Comment:
In these primitive encoding methods, we should specify the specific schema type, no? In this case `AvroNullSchema`.